src/Xamarin.Android.Build.Tasks/Utilities/ApplicationConfigNativeAssemblyGeneratorCLR.cs

```diff
@@ -0,0 +1,541 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+
+using Java.Interop.Tools.TypeNameMappings;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Xamarin.Android.Tasks.LLVMIR;
+
+namespace Xamarin.Android.Tasks;
+
+class ApplicationConfigNativeAssemblyGeneratorCLR : LlvmIrComposer
+{
+	// From host_runtime_contract.h in dotnet/runtime
+	const string HOST_PROPERTY_RUNTIME_CONTRACT = "HOST_RUNTIME_CONTRACT";
+	const string HOST_PROPERTY_BUNDLE_PROBE     = "BUNDLE_PROBE";
+	const string HOST_PROPERTY_PINVOKE_OVERRIDE = "PINVOKE_OVERRIDE";
+
+	sealed class DSOCacheEntryContextDataProvider : NativeAssemblerStructContextDataProvider
+	{
+		public override string GetComment (object data, string fieldName)
+		{
+			var dso_entry = EnsureType<DSOCacheEntry> (data);
+			if (String.Compare ("hash", fieldName, StringComparison.Ordinal) == 0) {
+				return $" from name: {dso_entry.HashedName}";
+			}
+
+			if (String.Compare ("name", fieldName, StringComparison.Ordinal) == 0) {
+				return $" name: {dso_entry.name}";
+			}
+
+			return String.Empty;
+		}
+	}
+
+	// Disable "Field 'X' is never assigned to, and will always have its default value Y"
+	// Classes below are used in native code generation, thus all the fields must be present
+	// but they aren't always assigned values (which is fine).
+	#pragma warning disable CS0649
+
+	// Order of fields and their type must correspond *exactly* (with exception of the
+	// ignored managed members) to that in
+	// src/monodroid/jni/xamarin-app.hh DSOCacheEntry structure
+	[NativeAssemblerStructContextDataProvider (typeof (DSOCacheEntryContextDataProvider))]
+	sealed class DSOCacheEntry
```

Comment:
Could these be `struct` rather than class? They might be more efficient that way?

Reply:
Sorry, I'm having a bit of a code optimisation phase 🤣

Reply:
It's OK, but there are plenty of those, can we leave it till after #9572 is merged? I don't want to make any more big changes - note that these files are copies of the existing MonoVM ones (with modifications, of course), so the modifications would have to be made to the originals too.